pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +525,165 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
+		klog.V(6).Infof("List Volumes in Cluster:")
+		return d.listVolumesInCluster(ctx, start, int(req.MaxEntries))
+	}
+	klog.V(6).Infof("List Volumes in Node Resource Group:")
+	return d.listVolumesInNodeResourceGroup(ctx, start, int(req.MaxEntries))
+}
 
-	disks, derr := d.cloud.DisksClient.ListByResourceGroup(ctx, d.cloud.ResourceGroup)
-	if derr != nil {
-		return nil, status.Error(codes.Internal, fmt.Sprintf("ListVolumes on rg(%s) failed with error: %v", d.cloud.ResourceGroup, derr.Error()))
+// listVolumesInCluster is a helper function for ListVolumes used for when there is an available kubeclient
+func (d *Driver) listVolumesInCluster(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
+	pvList, err := d.cloud.KubeClient.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "ListVolumes failed while fetching PersistentVolumes List with error: %v", err.Error())
+	}
+
+	// get all resource groups and put them into a sorted slice
+	rgMap := make(map[string]bool)
+	volSet := make(map[string]bool)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == DriverName {
+			diskURI := pv.Spec.CSI.VolumeHandle
+			if err := isValidDiskURI(diskURI); err != nil {
+				klog.Warningf("invalid disk uri (%s) with error(%v)", diskURI, err)
+				continue
+			}
+			rg, err := GetResourceGroupFromURI(diskURI)
+			if err != nil {
+				klog.Warningf("failed to get resource group from disk uri (%s) with error(%v)", diskURI, err)
+				continue
+			}
+			volSet[diskURI] = true
+			if _, visited := rgMap[rg]; visited {
+				continue
+			}
+			rgMap[rg] = true
+		}
 	}
 
-	if start != 0 && start >= len(disks) {
+	if start != 0 && start > len(volSet) {
 		return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
 	}
+	resourceGroups := make([]string, len(rgMap))
+	i := 0
+	for rg := range rgMap {
+		resourceGroups[i] = rg
+		i++
+	}
+	sort.Strings(resourceGroups)
 
-	maxEntries := len(disks) - start
-	if req.MaxEntries > 0 && int(req.MaxEntries) < maxEntries {
-		maxEntries = int(req.MaxEntries)
+	// loop through each resourceGroup to get disk lists
+	entries := []*csi.ListVolumesResponse_Entry{}
+	numVisited := 0
+	isCompleteRun := true
+	for _, resourceGroup := range resourceGroups {
+		if !isCompleteRun || (maxEntries > 0 && len(entries) >= maxEntries) {
+			isCompleteRun = false
+			break
+		}
+		listStatus := d.listVolumesByResourceGroup(ctx, resourceGroup, entries, start-numVisited, maxEntries-len(entries), &volSet)
+		numVisited += listStatus.numVisited
+		if listStatus.err != nil {
+			if status.Code(listStatus.err) == codes.FailedPrecondition {
+				continue
+			}
+			return nil, listStatus.err
+		}
+		entries = listStatus.entries
```

Comment:
Shouldn't we append the entries here instead of directly setting them. If we set it like this the entries from the previously iterated resource group will get overridden.

Reply:
entries get passed as a parameter to `listVolumesByResourceGroup` so the returned entries from `listVolumesByResourceGroup` will be always be appended entries including those from previously iterated resource group. So it shouldn't be a problem :)

Reply:
Oh ofcourse! My bad.